modules/core/src/webgl/classes/texture-cube.js

```diff
@@ -46,16 +50,18 @@ export default class TextureCube extends Texture {
     //   }
     // });
 
-    this.setCubeMapImageData(props);
-
-    // Called here so that GL.
-    // TODO - should genMipmap() be called on the cubemap or on the faces?
-    if (mipmaps) {
-      this.generateMipmap(props);
-    }
-
     // Store props for accessors
     this.opts = props;
+
+    this.setCubeMapImageData(props).then(() => {
+      this.loaded = true;
+
+      // TODO - should genMipmap() be called on the cubemap or on the faces?
+      // TODO - without generateMipmap() cube textures do not work at all!!! Why?
```

Comment:
Are you getting errors here? And does Luma use immutable textures? If you allocate storage beforehand and put aside space for mipmaps, then the texture would be incomplete before mipmaps are generated.

Reply:
> does Luma use immutable textures?

We haven't started using immutable texture, but it is definitely of interest, and the API should support it

> the texture would be incomplete before mipmaps are generated.

Good point. We do want an API that enables (and ideally makes it convenient) to work with such use cases.

Reply:
But still why does it not work if we don't call **generateMipmap**?

Reply:
@georgios-uber It's because the default value of MIN_FILTER is NEAREST_MIPMAP_LINEAR (https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/glTexParameter.xhtml).

If you try to draw with that min filter without generating the mip maps, the texture will be considered incomplete. 

The clue is following error being thrown in the console: `RENDER WARNING: texture bound to texture unit 0 is not renderable. It maybe non-power-of-2 and have incompatible texture filtering.`

Reply:
I'd recommend not allowing `generateMipmap` to be passed in as an argument, and instead deducing it from the min filter ([e.g.](https://github.com/tsherif/picogl.js/blob/master/src/texture.js#L114)).

Reply:
I tried changing the filter to LINEAR and did not fix the issue. Let me try again.

Reply:
BTW, are you using SpectorJS? https://spector.babylonjs.com/

It's a godsend for debugging stuff like this.